### PR TITLE
Adds full handling of permissions

### PR DESCRIPTION
**Pull request by AngellusMortis**
* Adds `Permission` model to parse permission strings from UniFi Protect
* Adds `can` method to user
* Adds `can_*` helper methods to Protect models to check permissions
* Splits all user related models into its own file, only the marked code sections are new for the `user.py`, the rest are moved from the old file.

Next PR will be to integrate the permissions into other locations/methods.

### CODE REVIEW

File: pyunifiprotect/data/user.py [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    ) -> str:
+        return self.raw_permission
+
+    @property
+    def obj(self) -> Optional[ProtectModel]:
+        if self.obj_id is None:
+            return None
+
+        devices = getattr(self.api.bootstrap, f"{self.model.value}s")
+        return cast(Optional[ProtectModel], devices.get(self.obj_id))
```
Comment (AngellusMortis) [inline pyunifiprotect/data/user.py:48]: New

File: pyunifiprotect/data/user.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    type: str
+    is_default: bool
+
+    @classmethod
+    def unifi_dict_to_dict(cls, data: Dict[str, Any]) -> Dict[str, Any]:
+        if "permissions" in data:
+            permissions = data.pop("permissions")
+            data["permissions"] = [{"rawPermission": p} for p in permissions]
+
+        return super().unifi_dict_to_dict(data)
```
Comment (AngellusMortis) [inline pyunifiprotect/data/user.py:63]: New

File: pyunifiprotect/data/user.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    @classmethod
+    def unifi_dict_to_dict(cls, data: Dict[str, Any]) -> Dict[str, Any]:
+        if "permissions" in data:
+            permissions = data.pop("permissions")
+            data["permissions"] = [{"rawPermission": p} for p in permissions]
+        if "allPermissions" in data:
+            permissions = data.pop("allPermissions")
+            data["allPermissions"] = [{"rawPermission": p} for p in permissions]
+
+        return super().unifi_dict_to_dict(data)
```
Comment (AngellusMortis) [inline pyunifiprotect/data/user.py:151]: New

File: pyunifiprotect/data/user.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        for perm in self.all_permissions:
+            if model == perm.model and node in perm.nodes:
+                if perm.obj_id is None:
+                    self._perm_cache[perm_str] = True
+                    return True
+                if perm.obj_id is not None and perm.obj == obj:
+                    self._perm_cache[perm_str] = True
+                    return True
+        self._perm_cache[perm_str] = False
+        return False
```
Comment (AngellusMortis) [inline pyunifiprotect/data/user.py:195]: New

File: pyunifiprotect/data/user.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            return self._perm_cache[perm_str]
+
+        for perm in self.all_permissions:
+            if model == perm.model and node in perm.nodes:
+                if perm.obj_id is None:
+                    self._perm_cache[perm_str] = True
+                    return True
+                if perm.obj_id is not None and perm.obj == obj:
+                    self._perm_cache[perm_str] = True
+                    return True
```
Comment (bdraco) [inline pyunifiprotect/data/user.py:193]: Usually I'd reverse and continue here to save some indent

**Review by bdraco — APPROVED** [review]

**Review by bdraco — APPROVED** [review]

**Pull request merged**